### ISSUE TITLE
[Experiment][MCP]: Evaluate schema-bearing direct operation tools against the fixed two-tool contract

### ISSUE THREAD

**Issue by morluto** [morluto/jacobian#1218]
## Summary

Jacobian currently represents hundreds of mathematical operations as data behind one MCP meta-tool:

```json
{
  "name": "math.run",
  "arguments": {
    "capability_id": "matrix.rank.compute",
    "payload": {}
  }
}
```

The selected operation’s actual input/output schemas, effects, execution behavior, and value type are not the schema of the tool the model calls. They were returned earlier by `math.find` as data. This erases the strongest native affordance MCP provides: every tool already has a unique name plus its own `inputSchema`, `outputSchema`, annotations, and execution metadata.

The erasure is the root of several compensating layers:

- the managed Skill teaches remembered IDs and payloads;
- #1031 proposes prepared invocation handles and host-specific temporary validators;
- `math.run` must be globally annotated for the most dangerous reachable operation;
- invalid nested payloads reach internal dispatch before capability-specific validation;
- the generic output schema is a universal result envelope rather than the operation’s typed result;
- packaged connectors can drift from canonical IDs/schemas while appearing to expose the same meta-tool.

Replace `math.run` with protocol-native schema-bearing mathematical tools. Keep `math.find` as semantic discovery if useful, but have it return canonical MCP tool names that are directly callable:

```text
math.find(query="exact matrix rank")
  -> matrix.rank.compute

matrix.rank.compute(matrix=...)
  -> typed rank result
```

No Skill, prepared-call token, generic payload wrapper, or compatibility alias should be required.

## MCP protocol evidence

The current MCP tool contract already defines each tool by:

- unique `name`;
- `inputSchema`;
- optional `outputSchema`;
- annotations;
- execution/task support.

`tools/list` is paginated, and current stateless guidance requires the tool set to be deterministic for a request’s authorization and **not** mutate per connection or as a side effect of earlier calls. This makes session-local “select a capability, then add a temporary tool” a poor final fit.

References:

- https://modelcontextprotocol.io/specification/draft/server/tools
- https://modelcontextprotocol.io/specification/draft/server/utilities/pagination
- https://modelcontextprotocol.io/specification/draft/changelog

The protocol-native representation of `matrix.rank.compute` is therefore a tool named `matrix.rank.compute` with the exact matrix request schema—not an enum/string inside another tool.

### The host sees an untyped executor

`capability_invoke()` advertises:

```python
capability_id: CapabilityId
payload: dict[str, Any]
```

The host can validate only the outer fields. It cannot constrain matrix shape, polynomial variables, verifier `input`/`candidate` siblings, discriminated unions, or artifact/reference compatibility before dispatch.

### `math.find` returns a type as data

Exact inspection can return the selected JSON Schema, but the agent/host must interpret and transplant that schema into a later generic call. #1031 accurately describes this as search-to-execution type erasure.

A prepared invocation handle can bind staleness, but it does not automatically make a standard MCP host use the selected schema for call generation. That requires host-specific integration—the same class of compatibility machinery the product should avoid.

### Effects and execution metadata are flattened

Because one `math.run` reaches pure arithmetic, artifact creation, durable experiments, and cancellation, its top-level annotation is conservatively destructive. Capability-specific effects/deadlines/cancellation cannot reach the standard tool boundary without selected-tool materialization or custom host logic.

### Results are flattened

The standard tool can publish a capability-specific `outputSchema`, but generic `math.run` instead publishes the universal `CapabilityResult` envelope. The agent must discover both the input and output type out of band.

### The Skill exists because actual functions are hidden

The managed Skill’s hardcoded direct-call catalog is effectively a hand-written function reference for operations that the MCP surface does not expose as actual functions. Removing the Skill under #1207 is easier when the live tool list contains the real operations and their schemas.

## Design principles from mature function libraries

Composable libraries expose concrete operations over stable typed values:

- Rust iterators share a tiny trait and associated item type; adapters are actual typed functions/methods.
- Scikit-learn pipelines chain objects implementing a concrete transformer interface.
- JAX transformations accept and return functions/pytrees through a shared typed structural protocol.
- MLIR operations expose generic interfaces without hiding every operation behind a stringly typed `run(op_name, payload)` dispatcher.

The common anti-pattern is a **stringly typed service locator/meta-dispatcher** that erases concrete function contracts.

References:

- https://doc.rust-lang.org/std/iter/
- https://scikit-learn.org/stable/modules/compose.html
- https://docs.jax.dev/en/latest/jax.tree_util.html
- https://mlir.llvm.org/docs/Interfaces/

### 1. `math.find` remains optional semantic discovery

```json
{"query":"exact rank over a finite field"}
```

It returns compact candidates containing canonical MCP tool names and typed applicability facts. Search does not execute or materialize temporary tools.

A caller that already knows `matrix.rank.compute` invokes it directly; no search sequence is mandatory.

### 2. Every installed mathematical operation is an MCP tool

Examples:

```text
integer.compute.gcd
matrix.rank.compute
polynomial.compute.gcd
finite_field.validate
linear_map.rank.compute
linear_map.rank.verify
```

Each definition is generated from the authoritative typed operation contract (#1212):

```text
name              = canonical capability ID
inputSchema       = exact request schema
outputSchema      = exact public result variant
annotations       = operation effect/idempotence
execution         = operation lifecycle/task support
title/description = compact domain-owned outcome
```

No `capability_id` argument and no generic `payload` wrapper.

### 3. `tools/list` is the callable inventory

Use standard pagination and deterministic ordering. `capability://catalog` may remain an operator/audit resource, but it is not a second callable registry.

The installed portfolio may still be policy/provider/auth scoped before request handling. It must not mutate because the agent searched or selected an operation.

### 4. Generate operation tools from one compiler

#1212’s typed operation compiler should generate the MCP binding directly. There should be no separately handwritten top-level tool schema per capability and no permanent wrapper that forwards into the old meta-dispatcher.

Runtime dispatch can still use one internal registry keyed by canonical ID. The public tool binding must preserve the selected request/result type rather than erasing it.

### 5. Compose typed values/references between real tools

#1208 ports and #1139 value references let one direct tool consume another tool’s exact output without model-authored JSON copying. The direct tool’s schema declares accepted inline/reference/artifact carriers.

### 6. Remove `math.run`

After a versioned cutover, delete:

- the public `math.run` tool;
- `capability_invoke()` as a public generic handler;
- `capability_id` + `payload` outer envelope;
- generic prepared-invocation tokens whose only purpose is restoring selected schema;
- global worst-case annotations;
- generic output-schema projection;
- Skill direct-run payloads;
- `capability.invoke` compatibility aliases.

An internal `registry.invoke(id, typed_request)` may remain an implementation detail and test seam.

## Tool-count and context evaluation

The strongest objection is the installed portfolio size. Evaluate it honestly against the meta-tool complexity rather than assuming two tool names means low context.

Measure on real supported hosts:

1. **Current meta-tool**
   - two top-level tools;
   - full generic input/output schemas;
   - server instructions/Skill;
   - repeated search/contract outputs;
   - invalid-call retries.
2. **Direct operation tools**
   - paginated deterministic `tools/list`;
   - actual host behavior for lazy/deferred tool loading, caching, and model injection;
   - no Skill and no repeated exact-inspection schemas.
3. **Domain/family grouping only as a measured fallback**
   - a small finite union per coherent family if a host demonstrably cannot handle direct tools;
   - never one universal untyped executor.

Record the complete model-facing context manifest from #1136. A host that fetches 350 definitions does not necessarily inject all 350 into every model turn; measure the actual binding/prompt behavior.

Required outcomes:

- valid first call;
- discovery-to-execution continuation;
- static and per-turn prompt bytes/tokens;
- call/result bytes;
- effect/confirmation correctness;
- latency;
- final mathematical correctness;
- irrelevant tool selection;
- cross-host consistency.

Do not reward the meta-tool merely for hiding schemas from `tools/list` if those schemas are then repeatedly returned through `math.find` and interpreted manually.

## Alternative if direct tools are impossible on one host

The fallback must remain schema-bearing and finite, for example a coherent domain-family tool with a closed discriminated input/output union generated from member operations. It must not accept an unconstrained payload object.

Host-specific limitations should be isolated in that host adapter and measured. They should not redefine the canonical server contract or require a Skill.

## Simplifications unlocked

- no generic execution meta-tool;
- no selected-schema-as-data impedance mismatch;
- no prepared invocation handle/state/version subsystem solely for validation;
- no `capability_id` and payload wrapper errors;
- truthful per-operation annotations and execution metadata;
- exact per-operation output schemas;
- direct standard client validation/autocomplete/code generation;
- no Skill fast-path catalog;
- simpler telemetry identity;
- simpler packaged connector generation;
- fewer discovery calls and smaller result-processing loops;
- `math.find` becomes search only, not part of a mandatory workflow.

## Acceptance criteria

- [ ] Installed mathematical operations are exposed as canonical schema-bearing MCP tools generated from typed operation contracts.
- [ ] The tool name is the canonical capability ID and its input/output schemas are validation-equivalent to runtime request/result types.
- [ ] `math.find` returns directly callable tool names and is optional for known operations.
- [ ] No standard client needs a Skill, prepared-call token, schema transplant, or generic payload wrapper.
- [ ] Operation-specific effects, idempotence, deadlines, cancellation/task support, and output types reach the native MCP tool definition.
- [ ] `tools/list` is deterministic, paginated, and consistent across stdio/HTTP/packaged hosts.
- [ ] Direct-tool context cost is measured from actual model-facing host projections and compared against the complete current meta-tool/Skill/search cost.
- [ ] Producer→consumer/checker composition preserves typed ports/value references.
- [ ] `math.run`, public `capability_invoke`, historical aliases, and generic prepared-invocation compatibility paths are removed after cutover.
- [ ] Runtime policy/provider/checker authorization remains authoritative and cannot be bypassed by seeing a tool definition.

## Non-goals

- Hand-writing hundreds of MCP bindings.
- Injecting the complete catalog into every prompt without measurement.
- Allowing the tool list to mutate per connection or after search.
- Exposing operator-only/internal storage functions as public math tools.
- Replacing typed mathematical values with one domain-family `payload: object`.
- Encoding research strategy or automatic workflow execution.

## Related

- #1029 — canonical packaged tool identity.
- #1031 — type erasure reproduction; this issue proposes removing the meta-tool rather than restoring schema through prepared handles.
- #1034 — capability-specific effects become native annotations.
- #1077 — capability-specific execution lifecycle becomes native tool metadata.
- #1136 — complete model-facing context measurement.
- #1207 — no-Skill contract.
- #1208 — typed operation ports/composition.
- #1209 — exact result variants.
- #1211 — simplified search/select; exact selection may reduce further to direct tool lookup.
- #1212 — typed operation compiler.
- #1214 — no compatibility shims.

**Comment by morluto** [comment]
## SDK-native simplification and remaining shim boundary

The exact MCP Python SDK v2.0.0 already performs the core machinery needed by direct operation tools:

```text
inspect typed Python callable
-> create Pydantic argument model
-> generate inputSchema
-> validate/convert arguments
-> invoke callable
-> validate/serialize typed structured output
-> generate outputSchema
```

Primary SDK sources:

- https://github.com/modelcontextprotocol/python-sdk/blob/v2.0.0/src/mcp/server/mcpserver/utilities/func_metadata.py
- https://github.com/modelcontextprotocol/python-sdk/blob/v2.0.0/src/mcp/server/mcpserver/tools/base.py
- https://github.com/modelcontextprotocol/python-sdk/blob/v2.0.0/docs/migration.md

This strengthens the deletion case for direct schema-bearing operation tools. Once the common typed operation compiler can expose a typed callable, the MCP boundary should contribute only:

```text
canonical tool name/title/description
operation-specific annotations/task support
runtime scope middleware
call into the compiled typed operation
```

It should not retain a parallel hierarchy of MCP projection models, hand-built output unions, a generic `payload: dict`, or a second result interpreter.

### Current SDK gap: strict unknown arguments

Jacobian's `JacobianMCPServer` subclass currently exists partly because SDK-generated argument models do not set `extra="forbid"`; unknown top-level arguments can otherwise be ignored. The subclass patches `additionalProperties: false` into `tools/list` and manually rejects undeclared arguments before delegating.

That is a real compatibility gap in the exact pinned SDK, but it should remain one tiny, documented temporary boundary—not grow into a custom server framework.

Preferred disposition:

1. open/upstream or adopt an SDK-supported strict-argument option;
2. pin a conformance fixture proving unknown fields fail;
3. after the SDK path exists, delete:
   - `JacobianMCPServer.list_tools()` schema rewriting;
   - `JacobianMCPServer.call_tool()` manual argument diffing;
   - the custom subclass if no other responsibility remains.

Do not solve strictness by preserving `math.run` and validating a generic inner payload manually. Direct operation tools need both:

```text
strict outer MCP arguments
+
authoritative typed mathematical request parsing
```

### Additional deletion targets enabled by direct tools

Current `adapters/mcp/tools.py` maintains transport-specific models for discovery and run results, manual text projections, generic result compaction, ResourceLink construction, and exact-inspection invocation reconstruction. Under direct operation bindings plus #1209/#1219:

- the operation's typed result variant is the structured output;
- text is one bounded renderer over that variant;
- explicit typed references produce ResourceLinks;
- SDK-generated output schema is authoritative;
- unknown-capability recovery belongs to `math.find`/tool-list lookup, not a generic executor result rewrite.

The direct-tool pilot should measure source deletion in the MCP adapter, not merely add generated direct bindings beside `math.run`.

**Comment by morluto** [comment]
## Cross-transport acceptance: the CLI must be generated from the same operation tool contract

#1225 found that Jacobian already has another schema-erasing execution surface besides `math.run`: the handwritten Typer CLI. It does not invoke the installed capability/operation boundary and can silently change lifecycle semantics.

If direct schema-bearing MCP tools become canonical, the same operation compiler should drive:

```text
MCP operation tool
CLI generic runner / selected generated convenience command
SDK/extension binding
```

A canonical operation ID must not mean one lifecycle over MCP and another over CLI. Current examples include `search.enumerate` returning a handle through the capability surface while `search-enumerate` starts and waits.

Add a direct-tool cross-transport gate:

```text
canonical typed request
-> operation tool
-> CLI operation runner
```

must agree on:

- request codec and contract digest;
- provider/policy/effect/lifecycle handling;
- mathematical value or checker verdict;
- artifact/record publication;
- diagnostics and recovery identity.

This also strengthens the no-Skill argument: generated CLI help/completion and standard MCP schemas are ordinary projections of the live contract. Neither needs copied payload guidance.

Do not preserve handwritten command aliases after the cutover under #1214. A small operator CLI can remain separate for state/snapshot/doctor/provider administration.

**Comment by morluto** [comment]
## Decision gate and implementation dependency update

This is now the preferred way to remove search→execution type erasure **if** the measured supported-host/tool-count gate passes. It should be generated from #1212's compiled operation contracts after #1226 removes operations that do not belong in the default toolbox.

### Do not expose the current catalog blindly

The direct callable inventory must be the post-disposition installed mathematical/checker portfolio, not today's complete generic workflow surface. Otherwise direct tools would merely project claim/plugin/experiment administration as hundreds of first-class operations.

### Required implementation order

```text
#1226 / #1229   delete or move obsolete workflow/reference operations
#1210 / #1212   establish authoritative typed operation codecs/compiler
#1209 / #1219   establish the closed result projection
#1218           generate native MCP tool bindings
#1207           delete any remaining Skill dependency
```

### Relationship to #1031

#1031 is now the fallback experiment for a concrete host limitation. Do not implement permanent prepared-invocation tokens, temporary tool state, and direct operation tools simultaneously.

### SDK ownership

Use the pinned MCP SDK's public typed tool registration, generated `inputSchema`, `outputSchema`, annotations, and execution metadata. Jacobian's current custom strict-extra-argument shim may remain only until the SDK or generated request model can express the same closed contract; it is not the foundation of per-operation binding generation.

### Acceptance addition

The implementation must prove:

- the tool list is deterministic for one authorization/catalog snapshot;
- pagination/order are stable and host-compatible;
- tool definitions are generated from the same request/result codecs used at execution;
- generic `math.run`, `capability_id + payload`, selected-call tokens, and Skill payload catalogs are deleted after cutover;
- actual prompt/model injection is measured on Codex, Claude Code, Cursor, stdio, and Streamable HTTP rather than inferred from MCP protocol size alone.

**Comment by morluto** [comment]
Closing as not planned under the retained product contract.

MCP SDK 2.0 makes direct schema-bearing operation tools technically feasible, but the repository has no measured host evidence that exposing the full operation catalog as top-level tools is simpler or more reliable than the fixed math.find / math.run surface. The immediate work is to delete the Skill and old product paths, preserve typed operation identity internally, and simplify selection/execution under #1207, #1210, #1211, #1212, and #1031.

Reconsider direct tools only if a future controlled host evaluation demonstrates that the fixed surface is the remaining bottleneck. Do not implement both architectures in parallel.

**Comment by morluto** [comment]
## Experimental decision gate: direct operation tools are a candidate, not the predetermined destination

The current fixed two-tool product contract is the control condition. This issue should compare it against schema-bearing direct operation tools using the **same surviving operation portfolio after #1226/#1229**.

A valid experiment may conclude that Jacobian should:

```text
KEEP_THE_FIXED_TWO_TOOL_CONTRACT
ADOPT_DIRECT_OPERATION_TOOLS
USE_A_BOUNDED_SCHEMA_BEARING_FAMILY_SURFACE
RUN_ANOTHER_HOST_SPECIFIC_EXPERIMENT
```

### Required host-level measurements

Measure the real model-facing behavior of supported hosts, not only server `tools/list` size:

- whether the host eagerly loads all tool definitions;
- how many definitions and bytes are injected into each model turn;
- caching/lazy-loading behavior;
- valid first execution-call rate;
- discovery-to-execution continuation;
- final mathematical task success;
- irrelevant tool selection;
- effect/confirmation correctness;
- latency and token/context cost;
- consistency across in-process, stdio, HTTP, packaged Codex, and installed connectors.

The two-tool control must include its complete costs—Skill/guidance if still present, repeated contract inspection, schema-transcription errors, and retries. The direct-tool arm must include actual host catalog loading and prompt projection.

### No premature cutover

Until the experiment meets a reviewed decision gate:

- do not make removal of `math.run` an acceptance criterion;
- do not build the full direct-tool generator for operations likely to be deleted by #1226;
- do not implement direct tools, prepared invocation, and family-union machinery simultaneously;
- do not treat protocol-native appearance as evidence of better agent performance.

The issue succeeds by producing a reproducible decision, even when the decision is to retain the fixed two-tool surface.
